### PR TITLE
Jupyterbook anyone?

### PR DESCRIPTION
**Pull request by tschm**
I have played with Jupyter Books which are imho somewhat nicer than Sphinx. Have a look please.
Also playing with your portfolio optimization in applications I have noticed that the construction

cp.sum_squares(np.transpose(np.linalg.cholesky(cov) @ w) is a lot faster than cp.quad form(w, cov) 

I am a bit puzzled here... @phschiele please comment.

The book you can build via

```bash
jupyter-book clean book
jupyter-book build book
```
Book is now book/_build/html. Push to a server you control...

### CODE REVIEW

**Comment by phschiele** [comment]
Hi @tschm,
thank you for this contribution! I was able to build the book with your instructions above, and it looks great!

Here is a preview: 
image

There are some external dependencies here, e.g., the [short course homepage](https://stanford.edu/~boyd/papers/cvx_short_course.html) points directly to paths on GitHub, so moving the notebooks would break these links. One option would be to host the HTML output on the GitHub pages web server of this repo and update the links on the short course website accordingly. A simple GitHub action can push the book to the `gh-pages` branch.

We'd have to check with the stakeholders of the short course (e.g., @stephenpboyd, @SteveDiamond).

Re your question about the portfolio optimization example:
If you profile the example, e.g. using `snakeviz`, you'll find that most time is spent in `is_psd_within_tol()`
image
(Due to caching, it looks like it is computed three times (3*1.31 > 1.49), but it's only computed once.)

The implicit PSD check in your version, i.e. computing the Cholesky factorization, is much faster but does not allow for the required tolerances. If you know that `P` in `quad_form` is PSD, e.g. because you checked beforehand, or because it is so by construction (like covariance matrices), @SteveDiamond added an `assume_PSD` argument in https://github.com/cvxpy/cvxpy/pull/1818. This feature will be available in CVXPY 1.3.

**Comment by tschm** [comment]
Thank you @phschiele. The book is still somewhat incomplete. The mbox is displayed in somewhat ugly font. Yes, it would be easy to link it to GitHub actions. Please note that all notebooks are executed (cache is an option here) and hence one could catch problems. Interesting insights on quad_form (thank you!) I also noticed, that not using the variable f and the constraint F.T w == f also results in a mild speed improvement. However, your main point is that the factor approach is much faster than using a full covariance shouldn't be polluted by quad_form issues. Hence I would use the faster Cholesky decomposition.

**Comment by tschm** [comment]
@phschiele I have to stop working on this for now but I have learnt more details both about Jupyter-Books but also about cvxpy. I have slightly rearranged the content of the repo and revisited the long README file.

This work is not finished though. It would need some integration via Github actions and a home somewhere on a server. 

There might be still a few issues with MathJax but they could be ironed out quickly.

**Comment by tschm** [comment]
@phschiele The approach taken here seems to be of cosmetic nature but there are features worth thinking about it. Any new notebook will pop up without any further required interaction in the TOC and will be linked from the central document. The book writes itself :-) You also have the search functionality across all notebooks and markdown files and you have the links to Binder, colab and a private JupyterHub server you maintain. The build of the book can be automated (through GitHub actions). Of course, one could go much deeper here and I point you to my inspiration:
https://python.quantecon.org/intro.html

**Comment by tschm** [comment]
I have revisited the problem of creating a jupyter book out of the content given in this repo. Let me emphasize that I have not touched the content but I have created the github workflows that take care of creating the book and caching the container used by Binder.

The book is served on https://tschm.github.io/cvx_short_course/docs/index.html. At compile time of the book all notebooks are executed and issues are reported as warnings. It's therefore a good way to control quality in notebooks.

The book also comes with a slick search functionality and you can search across all notebooks etc at once. Binder is integrated and hence students don't have to build a local environment at all. 

@stephenpboyd @phschiele @PTNobel @SteveDiamond

**Comment by tschm** [comment]
The only subtle change on all notebooks: The first cell has to be a Markdown cell with a # header. Otherwise the book would struggle to identify a string for the toc.

**Comment by stephenpboyd** [comment]
It’s awesome!

**Comment by tschm** [comment]
I don't have the powers to assign an official reviewer or to perform a merge request. Anyone keen?

**Comment by tschm** [comment]
The repo2docker construct is somewhat fragile and may not work at the moment. It's more robust to build an image and publish it on dockerhub. I have opened issues here: https://github.com/jupyterhub/repo2docker-action/issues/99

**Comment by tschm** [comment]
I am applying the docker hub path in tschm/antarctic

**Comment by tschm** [comment]
Please note that the construction of the image is not critical though. It reduces the time Binder needs to fire up a server as it can cache an image. However, this effect might be mitigated by many students using binder (and binder hence caching) or by accepting that it could take a while (e.g. 1 minute) to build the image as the cache is empty

File: .github/workflows/binder.yml
Comment (phschiele) [inline .github/workflows/binder.yml:5]: This comment refers to the other action.

**Review by phschiele — COMMENTED** [review]
@tschm Just have a few small comments on the PR above. Definitely, a much better experience looking at the book compared to the individual notebooks! The binder/colab integration makes it much more engaging.

It looks like some of the examples are missing subheadings.

Would your suggested course of action be to 
- merge this PR (after another review?)
- update the links that currently point to your repo back to `cvxgrp`
- update @stephenpboyd's [website](https://web.stanford.edu/~boyd/papers/cvx_short_course.html) to ensure that there are no broken links

File: .github/workflows/book.yml
Comment (tschm) [reply .github/workflows/book.yml]: https://github.com/cvxgrp/cvx_short_course/commit/2a2cd0c87e9b1dbdae51f788ed2d1fb33e1386d6

File: test.py
Comment (tschm) [reply test.py:1]: oh, no. This was more an accident. I will bring it back

Comment (tschm) [reply test.py]: https://github.com/cvxgrp/cvx_short_course/commit/7bda355c7080a966f680cf5d25ee91cea0d1bc8b

Comment (tschm) [reply test.py]: The subheadings should follow standard markdown conventions, e.g. the 
first cell should have a # header
The headers can't jump from # to ### etc. The number of # signs is not just an optical feature.

Comment (tschm) [reply test.py]: I don't think I removed any subheading btw. I removed or added "#" signs.

Comment (phschiele) [reply test.py]: @tschm I am not an admin for this repo. I talked to @PTNobel about the access rights, he'll give you an update in the lab.

File: .github/workflows/binder.yml
Comment (PTNobel) [inline .github/workflows/binder.yml:5]: How long does binder store a cached docker image? This repo is almost never updated and I'm not sure if the short course is taught often enough to justify this...

File: .github/workflows/book.yml
Comment (PTNobel) [inline .github/workflows/book.yml:27]: What does the `clean` step do?

File: README.md
Comment (PTNobel) [inline README.md:10]: Where did the other content from this file go?

File: book/_config.yml
Comment (PTNobel) [inline book/_config.yml:10]: I don't know how to read the double comment; does "force" or "auto" mean "only execute if there is at least on output cell missing"?

File: book/_config.yml
Comment (PTNobel) [inline book/_config.yml:34]: Please follow through on this comment

File: book/_config.yml
Comment (PTNobel) [inline book/_config.yml:36]: Should be `cvxgrp`

File: book/_config.yml
Comment (PTNobel) [inline book/_config.yml:52]: Why classic?

File: book/docs/applications/index.md
Comment (PTNobel) [inline book/docs/applications/index.md:1]: Should we include all the content from the README? Including what prereqs are expected?

File: book/docs/applications/notebooks/quantile_regression.ipynb
Comment (PTNobel) [inline book/docs/applications/notebooks/quantile_regression.ipynb:39]: I think this change violates Stephen's style guide

File: book/docs/conclusions.md
Comment (PTNobel) [inline book/docs/conclusions.md:1]: This seems like an outline for the instructor not the student; I thought the book was meant to be shown to students?

File: book/docs/constructive.md
Comment (PTNobel) [inline book/docs/constructive.md:1]: I don't follow this file; is this an outline of a new section?

File: book/docs/hello/notebooks/Lasso.ipynb
Comment (PTNobel) [inline book/docs/hello/notebooks/Lasso.ipynb:37]: Stephen's styleguide uses \begin{array} and \end{array} for optimization problems

**Review by PTNobel — CHANGES_REQUESTED** [review]
There's a number of comments and notes I had; overall this is great progress! I'm a bit confused on the README changes, it feels like you took an instructor facing part of the README and are now putting that content in student facing sections of the book? I'm not certain.

Is there somewhere I can look to see the whole thing rendered?

File: .github/workflows/binder.yml
Comment (tschm) [reply .github/workflows/binder.yml:5]: I don't know. This is very much an online learning experience :-) My limited personal experience is that the image stays cached at least for a few days. So I would fire off the server before I face the students....

File: .github/workflows/book.yml
Comment (tschm) [reply .github/workflows/book.yml:27]: It deletes the html files etc. generated in previous iterations. They may not be there and probably this could be deleted in a ci process

File: README.md
Comment (tschm) [reply README.md:10]: It's very much identical to the table of contents.

File: book/_config.yml
Comment (tschm) [reply book/_config.yml:10]: The force mode just means that all notebooks are executed from scratch not using any cache from previous runs. The auto is a different mode but I don't use it. Given we the time to build the book, force is the safest option

File: book/_config.yml
Comment (tschm) [reply book/_config.yml:36]: Yes, this should be changed first thing after the merge... otherwise the book fails on my repo. Did you have a look at https://tschm.github.io/cvx_short_course/docs/index.html?

File: book/docs/conclusions.md
Comment (tschm) [reply book/docs/conclusions.md:1]: Yes, it's for the students

**Pull request closed without merging**
